Create sonar images to track objects.c

Suruchibharti opened the pull request:
This PR adds a C program for generating, rotating, and smoothing an N×N sonar intensity matrix. The implementation adheres strictly to in-place operations using pointer arithmetic only, without using any extra N×N matrices or array indexing.  

Features Implemented
1:Matrix Generation:
Generates a random N×N matrix with values between 0 and 255.
Uses dynamic memory allocation and pointer arithmetic.

2: 90° Clockwise Matrix Rotation (In-Place)
Rotates the matrix by swapping elements through pointer references.
No additional memory or array indexing is used.
Rotation is achieved via transpose + row reversal logic.

3; Smoothing Filter 
Applies a smoothing filter that replaces each cell with the average of itself and its valid neighbors.
Handles boundary conditions correctly to avoid out-of-bound memory access.
Uses a temporary 1D pointer array for in-place smoothing (no secondary N×N matrix).

Suruchibharti committed:
Create sonar images to track objects.c

Initial commit

Suryanshsrivastv commented:
@Suruchibharti Please Add more descriptive PR title and description .

Suryanshsrivastv commented:
please check the input validation , the program works even when I enter "9a" .

Suryanshsrivastv commented on Assignment-3/sonar images to track objects.c:
```diff
@@ -0,0 +1,135 @@
+}
+
+// Rotate the matrix 90° clockwise in-place
+void rotateMatrix90Clockwise(int ***matrix, int size) {
```
could you please justify the use of ***(triple pointers here).

Suruchibharti replied:
> could you please justify the use of ***(triple pointers here).

@Suryanshsrivastv ,  I have used ***(triple pointers because the triple pointer int ***matrix is used here because the rotateMatrix90Clockwise function modifies the actual matrix pointer in-place.

In C, a single * pointer lets us modify the contents of the memory it points to.

A double pointer ** is typically used to point to an array of pointers (like our 2D matrix).

Here, *** allows us to pass the address of the 2D matrix pointer (int **matrix) so that any changes (like swapping rows via pointers) reflect in the original matrix in main.

 ***matrix gives us access to the 2D matrix itself through its pointer, which is necessary because we are performing in-place rotation using pointer arithmetic instead of copying values.
This approach avoids creating a temporary matrix and ensures the rotation directly updates the original matrix.

Suryanshsrivastv replied:
please try it with ** two pointers , i think this is also going to work , also i think it is more understandable , and I think the changes will be made in the original matrix(main) as long as we use pointers or pass the matrix by reference (which we do in case of pointers), and doesn't depend on the number of pointers used.

Suruchibharti replied:
> please try it with ** two pointers , i think this is also going to work , also i think it is more understandable , and I think the changes will be made in the original matrix(main) as long as we use pointers or pass the matrix by reference (which we do in case of pointers), and doesn't depend on the number of pointers used.

Hey @Suryanshsrivastv >,
you’re right — using two pointers would also work since the matrix is modified in-place through pointer references.
I used three pointers mainly for clarity and to handle edge rows safely, but I agree your approach would simplify it. Thanks for the suggestion!

Suruchibharti committed:
Update sonar images to track objects.c

input validation added

Suryanshsrivastv commented:
Hey @Suruchibharti , I was looking at the smoothing filter function and noticed something small. In the very first loop when row is 0, it seems like the prev row buffer is read from before it gets initialized. I think it might be using uninitialized memory for that first row's calculation. Just wanted to get your thoughts on it.

Suruchibharti committed:
Update sonar images to track objects.c

Updated rotateMatrix90Clockwise function to take matrix directly instead of its address

Suryanshsrivastv commented:
you could also use a bit more descriptive variable names like for ch,side,row,matrix etc.

Suruchibharti committed:
Update sonar images to track objects.c

initialize prevRow to prevent uninitialized memory access in smoothing filter and improve code clarity with descriptive variable names

Suruchibharti commented:
> Hey @Suruchibharti , I was looking at the smoothing filter function and noticed something small. In the very first loop when row is 0, it seems like the prev row buffer is read from before it gets initialized. I think it might be using uninitialized memory for that first row's calculation. Just wanted to get your thoughts on it.

I’ve fixed it by initializing prevRow with the first row’s values before the loop starts, so it no longer reads uninitialized memory. Thanks for pointing that out!

arti-niraj-yadav12 commented on Assignment-3/sonar images to track objects.c:
```diff
@@ -0,0 +1,147 @@
+
+// Generate a random N×N matrix with values 0–255
+void generateRandomMatrix(int **matrixData, int matrixSize) {
+    srand(time(0));
```
srand(time(0));  I THINK THIS  SHOULD BE IN MAIN FUNCTION
If this function is called multiple times within same second, it resets random seed and generates same sequence

arti-niraj-yadav12 commented on Assignment-3/sonar images to track objects.c:
```diff
@@ -0,0 +1,147 @@
+    int *currRow = (int *)malloc(matrixSize * sizeof(int));
+    int *nextRow = (int *)malloc(matrixSize * sizeof(int));
+
+    // Initialize prevRow and currRow with the first row’s values
```
Initializes BOTH prevRow and currRow with first row!
 This is incorrect - prevRow should represent row -1 (non-existent)
  For first row (rowIndex = 0), we should only use neighbors that exist
When processing row 0, column 0:
  - rowOffset = -1: neighborRow = -1 (should skip, but uses prevRow which has row 0 data!)
  - This causes incorrect averaging

arti-niraj-yadav12 commented on Assignment-3/sonar images to track objects.c:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    // Allocate memory for matrix
+    int **matrixData = (int **)malloc(matrixSize * sizeof(int *));
```
No null check after malloc?
your code should check ,if (matrixData == NULL) { error handling }

Suruchibharti committed:
Update sonar images to track objects.c

use srand(time(0)) in main function,
null check after malloc
fixed incorrect averaging and Uninitialized Reads .
